File: src/com/anathema_roguelike/main/ui/uielements/interactiveuielements/MenuScreen.java
/*******************************************************************************
 * Copyright (C) 2017 Carl Minden
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 ******************************************************************************/
package com.anathema_roguelike.main.ui.uielements.interactiveuielements;

import java.util.Collection;

import com.anathema_roguelike.main.ui.messages.Message;
import com.anathema_roguelike.main.ui.uielements.Screen;
import com.anathema_roguelike.main.ui.uielements.interactiveuielements.menus.Menu;
import com.anathema_roguelike.main.ui.uielements.interactiveuielements.menus.MenuDescription;
import com.anathema_roguelike.main.utilities.Utils;
import com.google.common.collect.Iterables;

public abstract class MenuScreen<T, M extends Menu<T>> extends Screen<T> {
	
	private M menu;
	private MenuDescription<T> description;
	private TextBox instructions;
	
	private Collection<T> choices;
	private Message instructionsMessage;
	private boolean cancellable;
	private float contentBackground;
	
	public MenuScreen(String title, Collection<T> choices, boolean cancellable) {
		super(title, cancellable);
		init(choices, null, cancellable, 1f);
	}
	
	public MenuScreen(String title, Collection<T> choices, Message instructionsMessage, boolean cancellable) {
		super(title, cancellable);
		init(choices, instructionsMessage, cancellable, 1f);
	}

	public MenuScreen(int x, int y, int width, int height, String title, Message instructionsMessage,
			boolean cancellable, float background, float contentBackground, Collection<T> choices) {
		
		super(x, y, width, height, title, cancellable, background);
		init(choices, instructionsMessage, cancellable, contentBackground);
	}
	
	public MenuScreen(int x, int y, int width, int height, String title, boolean cancellable, float background,
			float contentBackground, Collection<T> choices) {
		
		super(x, y, width, height, title, cancellable, background);
		init(choices, null, cancellable, contentBackground);
	}
	
	@Override
	public T run() {
		int x = getX() + (int)(getWidth() * .05);
		int y = getY() + (int)(getHeight() * .2);
		
		menu = createMenu(x, y, (int)(getWidth() * .3), getHeight()/2, choices, cancellable, contentBackground);
		description = new MenuDescription<T>(menu, choices.isEmpty() ? null : Utils.getSuperclass(Iterables.get(choices, 0)));
		
		if(instructionsMessage != null) {
			instructions = new TextBox(x, y - 2, menu.getWidth() + description.getWidth(), 1, instructionsMessage, contentBackground);
			addUIElement(instructions);
		}
		
		addUIElement(description);
		setFocusedUIElement(menu);
		
		return super.run();
	}
	private void init(Collection<T> choices, Message instructionsMessage, boolean cancellable, float contentBackground) {
		this.choices = choices;
		this.instructionsMessage = instructionsMessage;
		this.cancellable = cancellable;
		this.contentBackground = contentBackground;
	}
	
	public Collection<T> getChoices() {
		return choices;
	}
	
	protected void setInstructionsMessage(Message message) {
		instructions.setMessage(message);
	}
	
	public M getMenu() {
		return menu;
	}
	
	protected abstract M createMenu(int x, int y, int width, int height, Collection<T> choices, boolean cancellable, float background);
}

File: src/com/anathema_roguelike/main/ui/uielements/interactiveuielements/MenuScreen.scala
/*******************************************************************************
 * Copyright (c) 2019. Carl Minden
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 ******************************************************************************/

package com.anathema_roguelike
package main.ui.uielements.interactiveuielements

import com.anathema_roguelike.main.ui.UIConfig
import com.anathema_roguelike.main.ui.messages.Message
import com.anathema_roguelike.main.ui.uielements.Screen
import com.anathema_roguelike.main.ui.uielements.interactiveuielements.menus.{AbstractMenu, MenuDescription}
import com.anathema_roguelike.main.utilities.Utils
import com.anathema_roguelike.main.utilities.position.{Direction, Point}

abstract class MenuScreen[T, M <: AbstractMenu[T]](
    title: String,
    choices: Array[_ <: T],
    position: Point = new Point(0, 0),
    width: Int = UIConfig.TERM_WIDTH,
    height: Int = UIConfig.TERM_HEIGHT,
    background: Float = 1f,
    contentBackground: Float = 1f,
    cancellable: Boolean = true,
    var instructionsMessage: Option[Message] = None,
  ) extends Screen[T](title, cancellable, position, width, height, background) {

  private val x: Int = getX + (getWidth * .05).toInt
  private val y: Int = getY + (getHeight * .2).toInt

  private var instructions: Option[TextBox] = None
  private var menu: Option[M] = None

  override def run: Option[T] = {

    val menu = createMenu(Point(x, y), (getWidth * .3).toInt, getHeight / 2, choices, cancellable, contentBackground)

    val description = new MenuDescription[T](menu, choices.headOption.map(c => Utils.getSuperclass(c)))

    instructionsMessage.foreach(msg => {
      val i = new TextBox(Direction.offset(getPosition, Direction.DOWN, 2), menu.getWidth + description.getWidth, 1, contentBackground, msg)

      instructions = i
      addUIElement(i)
    })

    addUIElement(description)
    setFocusedUIElement(menu)
    super.run
  }

  protected def getMenu: Option[M] = menu

  protected def setInstructionsMessage(message: Message): Unit = {
    instructions.get.setMessage(message)
  }

  def getChoices: Iterable[_ <: T] = choices

  protected def createMenu(position: Point, width: Int, height: Int, choices: Array[_ <: T], cancellable: Boolean, background: Float): M
}
